swarm/chunk/tags.go

```diff
@@ -0,0 +1,233 @@
+		name:      s,
+		startedAt: time.Now(),
+		total:     uint32(total),
+		State:     make(chan State, 5),
+	}
+	_, loaded := ts.tags.LoadOrStore(s, t)
+	if loaded {
+		return nil, errExists
+	}
+	return t, nil
+}
+
+// Inc increments the state count for a tag if tag is found
+func (ts *tags) Inc(s string, f State) {
+	t, ok := ts.tags.Load(s)
+	if !ok {
+		return
+	}
+	t.(*Tag).Inc(f)
+}
+
+// Get returns the state count for a tag
+func (ts *tags) Get(s string, f State) int {
+	t, _ := ts.tags.Load(s)
+	return t.(*Tag).Get(f)
+}
+
+// State is the enum type for chunk states
+type State = uint32
+
+const (
+	SPLIT  State = iota // chunk has been processed by filehasher/swarm safe call
+	STORED              // chunk stored locally
+	SENT                // chunk sent to neighbourhood
+	SYNCED              // proof is received; chunk removed from sync db; chunk is available everywhere
+)
+
+// Tag represents info on the status of new chunks
+type Tag struct {
+	uid       uint32     // a unique identifier for this tag
+	name      string     // a name tag for this tag
+	total     uint32     // total chunks belonging to a tag
+	split     uint32     // number of chunks already processed by splitter for hashing
+	stored    uint32     // number of chunks already stored locally
+	sent      uint32     // number of chunks sent for push syncing
+	synced    uint32     // number of chunks synced with proof
+	startedAt time.Time  // tag started to calculate ETA
+	State     chan State // channel to signal completion
+}
+
+// New creates a new tag, stores it by the name and returns it
+// it returns an error if the tag with this name already exists
+func NewTag(uid uint32, s string, total uint32) *Tag {
+	if len(s) == 0 {
+		s = fmt.Sprintf("upload_%d", time.Now().Unix())
+	}
+	t := &Tag{
+		uid:       uid,
+		name:      s,
+		startedAt: time.Now(),
+		total:     uint32(total),
+		State:     make(chan State, 5),
+	}
+	return t
+}
+
+// Inc increments the count for a state
+func (t *Tag) Inc(state State) {
+	var v *uint32
+	switch state {
+	case SPLIT:
+		v = &t.split
+	case STORED:
+		v = &t.stored
+	case SENT:
+		v = &t.sent
+	case SYNCED:
+		v = &t.synced
+	}
+	n := atomic.AddUint32(v, 1)
+	if int(n) == t.GetTotal() {
+		t.State <- state
+	}
+}
+
+// Get returns the count for a state on a tag
+func (t *Tag) Get(state State) int {
+	var v *uint32
+	switch state {
+	case SPLIT:
+		v = &t.split
+	case STORED:
+		v = &t.stored
+	case SENT:
+		v = &t.sent
+	case SYNCED:
+		v = &t.synced
+	}
+	return int(atomic.LoadUint32(v))
+}
+
+// GetUid returns the unique identifier
+func (t Tag) GetUid() uint32 {
+	return t.uid
+}
+
+func (t Tag) GetName() string {
+	return t.name
+}
+
+// GetTotal returns the total count
+func (t *Tag) GetTotal() int {
+	return int(atomic.LoadUint32(&t.total))
+}
+
+// SetTotal sets total count to SPLIT count
+// is meant to be called when splitter finishes for input streams of unknown size
+func (t *Tag) SetTotal() int {
```

Comment:
i'm not sure why we need this @zelig. shouldn't the chunker increment the total on-the-fly? otherwise your comment is correct @janos

Reply:
if at tag creation time we know the file size, than we should set it, cos otherwise no progress bar for split can be given.

`SetTotal` poses no problem if it is out of sync at all. 
However, this is meant to be called when splitting finished. 
so better call it `DoneSplitting`

Reply:
@zelig you know the file size of just a single file at tag creation. you don't know the overall size of a directory because it's a tar stream. also, you know the size of the octet stream but you have no clue about the size of the manifest the encloses it; this also changes when you have entries that fall under same same node in the trie structure.
i vote to relieve ourselves from the notion that we know the size of a tag a-priori. i claim that for the most of it i think we don't (on the syncing node) and that implementing this would be prone to bugs and test flakiness. better just increment the counters from the context. why do we have to complicate this?

Reply:
i would hate to lose proper progress bar or ETA for the simplest case of uploading one large file.
In fact i would not even worry about the extra chunks of the wrapping manifest  at all.

Reply:
Have it your way. I rather develop features that cover every case rather than just a special case.